lib/process/process.go

```diff
@@ -718,49 +723,74 @@ func (p *Process) reconcileNodeLabels(ctx context.Context, client *kubernetes.Cl
 		return trace.Wrap(err)
 	}
 	for ip, node := range nodes {
-		if err := p.reconcileNode(client, *cluster, ip, node); err != nil {
+		if err := p.reconcileNode(ctx, client, *cluster, ip, node); err != nil {
 			p.WithError(err).Errorf("Failed to reconcile labels for node %v/%v.",
 				node.Name, ip)
 		}
 	}
 	return nil
 }
 
-func (p *Process) reconcileNode(client *kubernetes.Clientset, cluster ops.Site, ip string, node v1.Node) error {
+func (p *Process) reconcileNode(ctx context.Context, client *kubernetes.Clientset, cluster ops.Site, ip string, node v1.Node) error {
 	server, err := cluster.ClusterState.FindServerByIP(ip)
 	if err != nil {
 		return trace.Wrap(err)
 	}
-	missingLabels, err := getMissingLabels(cluster, *server, node)
+	profile, err := cluster.App.Manifest.NodeProfiles.ByName(server.Role)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	oldData, err := json.Marshal(node)
 	if err != nil {
 		return trace.Wrap(err)
 	}
-	if len(missingLabels) == 0 {
+
+	requiredLabels := server.GetNodeLabels(profile.Labels)
+	needUpdate := false
+	node.Labels, needUpdate = reconcileLabels(p, node.Labels, requiredLabels)
+	if needUpdate {
 		return nil
 	}
-	p.Infof("Adding missing labels to node %v/%v: %v.", node.Name, ip, missingLabels)
-	for key, val := range missingLabels {
-		node.Labels[key] = val
+	newData, err := json.Marshal(node)
+	if err != nil {
+		return trace.Wrap(err)
 	}
-	if _, err := client.CoreV1().Nodes().Update(context.TODO(), &node, metav1.UpdateOptions{}); err != nil {
+	patchData, err := jsonpatch.CreateMergePatch(oldData, newData)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	p.Infof("Patching node %v/%v: %s", node.Name, ip, string(patchData))
+	if _, err := client.CoreV1().Nodes().Patch(ctx, node.Name, types.MergePatchType, patchData, metav1.PatchOptions{}); err != nil {
 		return rigging.ConvertError(err)
 	}
 	return nil
 }
 
-func getMissingLabels(cluster ops.Site, server storage.Server, node v1.Node) (map[string]string, error) {
-	profile, err := cluster.App.Manifest.NodeProfiles.ByName(server.Role)
+func reconcileLabels(logger logrus.FieldLogger, currentLabels, requiredLabels map[string]string) (map[string]string, bool) {
+	labels := make(map[string]string)
+	for key, value := range currentLabels {
+		labels[key] = value
+	}
+	needUpdate := false
+	reconcileMode, err := defaults.ParseReconcileMode(labels[defaults.KubernetesReconcileLabel])
 	if err != nil {
-		return nil, trace.Wrap(err)
+		logger.WithError(err).Error("Unable to get reconcile mode, will reset to EnsureExists.")
```

Comment:
Because this is primarily a troubleshooting tool, I'm curious about the tradeoffs of this approach vs treating a missing label as enabled. 

I suspect this approach is slightly easier to use, since someone can just look at the label to find the right label and edit the mode. On the other hand it's another label that under normal circumstances wouldn't really need to be set since it's really for troubleshooting.

In the wild too, I could see some of the on-prem teams struggling to use this a bit, which might lead to some cli options to set/remove the label / show the status. In which case the label could just be missing since the main UX would be a CLI helper. 

I think both approaches are valid, so if you want to stick with this approach I'm fine with it.

Reply:
In my opinion, if the label is missing, it should be added automatically. This helps the operator team not to look for the label name in the documentation, but simply execute: `kubectl get no --show-labels`

Reply:
Yea, makes sense.